### PR TITLE
Update haskell-style.md

### PR DESCRIPTION
**Pull request by cdparks**
Add some more examples and counter-examples. Am I missing anything?

### CODE REVIEW

File: haskell-style.md [hunk 1]
```diff
@@ -155,3 +155,167 @@ instance FromJSON (GameSession Never MathStandardAssignmentId) where
...
+
+#### Data declarations
+```haskell
+--- Bad
+data SomeRecord = SomeRecord { someField :: Int
+                             , someOtherField :: Double
+                             } deriving (Eq, Show)
+
+-- Good
+data SomeRecord' =
```
Comment (pseudonom) [inline haskell-style.md]: `=` on the next line please.

File: haskell-style.md [hunk 2]
```diff
@@ -155,3 +155,167 @@ instance FromJSON (GameSession Never MathStandardAssignmentId) where
...
+             | SecondConstructor Int
+             | ThirdConstructor Double Text
+               deriving (Eq, Show)
+
+-- Good
+data SomeSum'
+  = FirstConstructor'
+  | SecondConstructor' Int
+  | ThirdConstructor' Double Text
+    deriving (Eq, Show)
```
Comment (pseudonom) [inline haskell-style.md]: I usually put `deriving` directly under the `|`.

File: haskell-style.md [hunk 3]
```diff
@@ -155,3 +155,167 @@ instance FromJSON (GameSession Never MathStandardAssignmentId) where
...
+#### Let expressions
+```haskell
+-- Bad - aligning multiple let bindings like this is fiddly
+someBinding mx =
+  let ma = fmap (+1) mx
+      mb = fmap (*2) mx
+  in (+) <$> ma <*> mb
+
+-- Good - put let on its own line and then we can use normal spacing
+someBinding' mx =
```
Comment (pseudonom) [inline haskell-style.md]: We can get rid of this one please.

File: haskell-style.md [hunk 4]
```diff
@@ -155,3 +155,167 @@ instance FromJSON (GameSession Never MathStandardAssignmentId) where
...
+someBinding mx =
+  f <$> mx <*> mx
+ where
+  f x y = x * x + y * y
+
+-- This style is more in the spirit of context-free alignment
+someBinding' mx =
+  f <$> mx <*> mx
+    where
+      f x y = x * x + y * y
```
Comment (pseudonom) [inline haskell-style.md:269]: Maybe add `someBinding mx = f <$> mx <*> mx where f x y = x * x + y * y`?

File: haskell-style.md [hunk 5]
```diff
@@ -155,3 +155,167 @@ instance FromJSON (GameSession Never MathStandardAssignmentId) where
...
+   where -- PARSE ERROR - parser thinks this where is a part of the last expression
+    f x y = x * x + y * y
+
+-- Instead, you have to do either this
+someBinding''' mx my = do
+  x <- mx
+  y <- my
+  f x y
+ where
+   f x y = x * x + y * y
```
Comment (pseudonom) [inline haskell-style.md]: 2 - 1 - 2 here please.

File: haskell-style.md [hunk 6]
```diff
@@ -155,3 +155,167 @@ instance FromJSON (GameSession Never MathStandardAssignmentId) where
...
+
+-- or this
+someBinding'''' mx my = do
```
Comment (pseudonom) [inline haskell-style.md]: Get rid of this please.

**Comment by pseudonom** [comment]
Yay, thank you!

**Comment by akurilin** [comment]
This is cool.

File: haskell-style.md (hunk 1 above)
Comment (cdparks) [reply haskell-style.md]: Done

File: haskell-style.md (hunk 2 above)
Comment (cdparks) [reply haskell-style.md]: Done

File: haskell-style.md (hunk 3 above)
Comment (cdparks) [reply haskell-style.md]: Done

File: haskell-style.md (hunk 4 above)
Comment (cdparks) [reply haskell-style.md:269]: Done

File: haskell-style.md (hunk 5 above)
Comment (cdparks) [reply haskell-style.md]: Done

File: haskell-style.md (hunk 6 above)
Comment (cdparks) [reply haskell-style.md]: Done

**Comment by cdparks** [comment]
Yeah, I did it all from github too. Wild!

**Comment by pseudonom** [comment]
We live in the future!

**Pull request merged**